pcdbapi/main.py

```diff
@@ -0,0 +1,160 @@
+import asyncio
+import os
+import secrets
+from contextlib import asynccontextmanager
+from datetime import datetime
+from typing import List, Literal, Optional
+
+from bson import ObjectId
+from fastapi import FastAPI, HTTPException, Request
+from motor.motor_asyncio import AsyncIOMotorClient
+from pydantic import BaseModel, ConfigDict, HttpUrl, constr
+
+from middleware import encode_response, verify_signature
+
+# All config in one place
+CONFIG = {
+    "mongodb_url": os.getenv("MONGODB_URL"),
+    "api_secret": os.getenv("API_SECRET"),  # shared secret
+    "max_timestamp_diff": 300,  # 5 minutes
+}
+
+
+class OidcClient(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    # These are the only fields that can be set through the API
+    name: constr(min_length=1, max_length=200) | None = None
+    redirect_uris: List[str] | None = None
+    post_logout_redirect_uris: List[str] | None = None
+    id_token_signed_response_alg: Literal["RS256", "ES256", "HS256"] | None = None
+    userinfo_signed_response_alg: Literal["RS256", "ES256", "HS256"] | None = None
+    active: bool | None = None
+
+
+def validate_objectid(_id: str) -> ObjectId:
+    try:
+        oid = ObjectId(_id)
+    except:
+        raise HTTPException(status_code=422, detail="Invalid ObjectId") from None
+    return oid
+
+
+@asynccontextmanager
+async def lifespan(app: FastAPI):
+    # Startup
+    app.mongodb_client = AsyncIOMotorClient(CONFIG["mongodb_url"])
```

Comment:
A TLS certificate must be provided

```suggestion
    app.mongodb_client = AsyncIOMotorClient(
        CONFIG["mongodb_url"],
        tls=True,
        tlsCertificateKeyFile='/path/to/client.pem',
        tlsCAFile='/path/to/ca.pem')
```